File: app/models/product.rb
class Product < ActiveRecord::Base
  include Concerns::CompanyAsset

  has_and_belongs_to_many :categories

  belongs_to :article,
      :polymorphic => true,
      :dependent   => :destroy

  has_many :rental_periods

  has_many :buying_prices,
      :class_name => 'ProductBuyingPrice',
      :dependent => :destroy

  has_many :distributors,
      :through => :buying_prices

  accepts_nested_attributes_for :buying_prices,
      :allow_destroy => true,
      :reject_if => lambda {|bp| bp[:distributor_id].blank? }

  delegate :full_name,
      :unit,
      :vat_rate,
      :available_count,
      :rental_price,
      :selling_price,
      :to => :article

  scope :matching, lambda {|str| where([%q(CONCAT(products.name, '-', products.code, '-', products.description) REGEXP ?), str])}

  scope :available_between, lambda {|from_date, to_date|
    where [%q(products.article_id NOT IN (SELECT product_stock_entries.device_id FROM product_stock_entries
              JOIN unavailabilities ON unavailabilities.product_stock_entry_id = product_stock_entries.id
              JOIN rental_periods   ON rental_periods.id = unavailabilities.rental_period_id
              JOIN rentals          ON rentals.id = rental_periods.rental_id
              WHERE (rentals.end NOT BETWEEN ? AND ?) AND (rentals.begin  NOT BETWEEN ? AND ?))), from_date, to_date, from_date, to_date]
  }

  scope :service, where(:article_type => 'Service')

  scope :device, where(:article_type => 'Device')

  validates_presence_of :article,
      :company_section

  collects_errors_from :article, :distributors, :buying_prices

  def method_missing(method_name, *attrs)
    if self.article and self.article.respond_to? method_name
      self.article.send method_name, *attrs
    else
      super
    end
  end

  def is_rentable?
    self.article.respond_to? :is_rentable? and self.article.is_rentable?
  end

  def is_sellable?
    self.article.respond_to? :is_sellable? and self.article.is_sellable?
  end

  def service?
    self.article.is_a? Service
  end

  def to_json(options={})
    super :include => :article
  end

  def availability(from, to)
    self.article.respond_to?(:availability) ? self.article.availability(from, to) : 1
  end

  def article_type=(type)
    raise 'article is already initialized' unless self.article.nil?
    case type
    when 'Device'
      self.article = Device.new
    when 'Service'
      self.article = Service.new
    else
      raise "unknown article_type: #{type}"
    end
  end

  def article_type
    self.article ? self.article.class.to_s : nil
  end

  def device_attributes=(attrs)
    self.article_type ||= 'Device'
    self.article.attributes = attrs
  end

  def service_attributes=(attrs)
    self.article_type ||= 'Service'
    self.article.attributes = attrs
  end

end

File: public/javascripts/product.js
var Product = ActiveResource.inherit({
  has_many: {
    'rental_periods': {}
  },
  belongs_to: {
    'company_section': {}
  },
  resource: 'products',

  isAvailable: function(from_date, to_date) {
    if (typeof(from_date) != 'object')
      from_date = new Date(from_date);
    if (typeof(to_date) != 'object')
      to_date = new Date(to_date);

    var unavailabilities = this.rental_periods();
    for (var i=0, len=unavailabilities.length; i<len; i++) {
      unav = unavailabilities[i];
      if (unav.from_date < to_date && unav.to_date > from_date) return false;
    }
    return true;
  },

  unit_price: function() {
    if (this.is_rentable())
      return $M(this.get('article').rental_price);
    else if (this.is_sellable())
      return $M(this.get('article').selling_price);
    else if (this.is_service())
      return $M(this.get('article').unit_price);
    else
      return $M(0);
  },

  rental_price: function() {
    return $M(this.get('article').rental_price || 0);
  },

  selling_price: function() {
    return $M(this.get('article').selling_price || 0);
  },

  is_rentable: function() {
    return !!this.get('article').rental_price;
  },

  is_sellable: function() {
    return !!this.get('article').selling_price;
  },

  is_service: function() {
    return this.get('article_type') == 'Service'
  }

});
